### PR TITLE
Replacing openid extension with OAuth2-proxy

### PR DESCRIPTION
**Pull request by eladiw**
#1386 #1110 

## What is being addressed

Since openid extension only returned the token_id and we needed the access token, we are changing the way we authenticate:
use oauth2 proxy instead of the guacamole open id extension.
This also fix the issue of not supporting refresh tokens as the proxy handles this

## How is this addressed

- add a service to the tomcat server that starts up the oauth2-proxy
- This service is the new entry point to the web app
- proxy listens on 8085 port and redirect traffic after its being authenticated
- our custom guacamole extension gets the token as a request header and pass it if needed to the underlying api
- extra: upgraded guacamole to 1.4.0

## Why is this a **_draft_** PR?
-- UPDATED --

- [x]  tests are not finished - Decided to do in the next iteration
- [x] port is 8085 but it should be 8080 -Decided to leave it like that
- [x]  need to validate the refresh token more thoroughly (tested manually over 9 hours with an open session)
- [x]  option to consider: not using tomcat server and build another docker file that hosts the needed processes - decided not to go on with that option.
- [x] vm list is not updated without logging out and in (regression) - will be handled separately
- [x] When you login to Guacamole you get the login screen (which can be configured to be skipped):
image

### CODE REVIEW

File: templates/workspace_services/guacamole/guacamole-server/docker/Dockerfile [hunk 1]
```diff
@@ -53,14 +53,12 @@ COPY ./docker/services /etc/services.d/
...
 
-RUN rm -rf ./guacamole-auth-openid-1.3.0
-
-RUN wget -O ${GUACAMOLE_HOME}/guacamole.war 'http://apache.org/dyn/closer.cgi?action=download&filename=guacamole/1.3.0/binary/guacamole-1.3.0.war'
+RUN wget -O  /etc/oauth2/oauth2-proxy-v7.2.1.linux-amd64.tar.gz 'https://github.com/oauth2-proxy/oauth2-proxy/releases/download/v7.2.1/oauth2-proxy-v7.2.1.linux-amd64.tar.gz'
 
 ENTRYPOINT [ "/init" ]
-EXPOSE 80 2222
+
+EXPOSE 80 2222 443
```
Comment (eladiw) [inline templates/workspace_services/guacamole/guacamole-server/docker/Dockerfile:64]: I think the 443 change can be removed

File: templates/workspace_services/guacamole/terraform/web_app.tf [hunk 2]
```diff
@@ -35,15 +35,15 @@ resource "azurerm_app_service" "guacamole" {
...
-    OPENID_AUTHORIZATION_ENDPOINT = "https://login.microsoftonline.com/${local.aad_tenant_id}/oauth2/v2.0/authorize"
-    OPENID_JWKS_ENDPOINT          = "https://login.microsoftonline.com/${local.aad_tenant_id}/discovery/v2.0/keys"
-    OPENID_ISSUER                 = "https://login.microsoftonline.com/${local.aad_tenant_id}/v2.0"
-    OPENID_CLIENT_ID              = "${var.openid_client_id}"
-    OPENID_REDIRECT_URI           = "https://${local.webapp_name}.azurewebsites.net/guacamole/"
-    OPENID_USERNAME_CLAIM_TYPE    = "email"
+    OAUTH2_PROXY_CLIENT_ID       = "${var.ws_client_id}"
+    OAUTH2_PROXY_CLIENT_SECRET   = "${var.ws_client_secret}"
+    OAUTH2_PROXY_REDIRECT_URI    = "https://${local.webapp_name}.azurewebsites.net/oauth2/callback"
+    OAUTH2_PROXY_EMAIL_DOMAIN    = "\"*\""
```
Comment (eladiw) [inline templates/workspace_services/guacamole/terraform/web_app.tf:44]: please add a comment why you escaped the "  chars

File: templates/workspace_services/guacamole/guacamole-server/docker/Dockerfile (hunk 1 above)
Comment (anatbal) [reply templates/workspace_services/guacamole/guacamole-server/docker/Dockerfile]: Fixed

File: templates/workspace_services/guacamole/terraform/web_app.tf (hunk 2 above)
Comment (anatbal) [reply templates/workspace_services/guacamole/terraform/web_app.tf]: Added

File: templates/workspace_services/guacamole/terraform/web_app.tf [hunk 3]
```diff
@@ -35,15 +35,15 @@ resource "azurerm_app_service" "guacamole" {
...
     ISSUER                = local.issuer
 
-    OPENID_AUTHORIZATION_ENDPOINT = "https://login.microsoftonline.com/${local.aad_tenant_id}/oauth2/v2.0/authorize"
-    OPENID_JWKS_ENDPOINT          = "https://login.microsoftonline.com/${local.aad_tenant_id}/discovery/v2.0/keys"
-    OPENID_ISSUER                 = "https://login.microsoftonline.com/${local.aad_tenant_id}/v2.0"
-    OPENID_CLIENT_ID              = "${var.openid_client_id}"
-    OPENID_REDIRECT_URI           = "https://${local.webapp_name}.azurewebsites.net/guacamole/"
-    OPENID_USERNAME_CLAIM_TYPE    = "email"
+    OAUTH2_PROXY_CLIENT_ID       = "${var.ws_client_id}"
+    OAUTH2_PROXY_CLIENT_SECRET   = "${var.ws_client_secret}"
```
Comment (marrobi) [inline templates/workspace_services/guacamole/terraform/web_app.tf:42]: Would be nicer if this was stored in KeyVault as part of the workspace deployment and the KV secret referenced. Maybe one to pick up in a separate issue?

File: templates/workspace_services/guacamole/terraform/web_app.tf [hunk 4]
```diff
@@ -35,15 +35,15 @@ resource "azurerm_app_service" "guacamole" {
...
+    OAUTH2_PROXY_REDIRECT_URI    = "https://${local.webapp_name}.azurewebsites.net/oauth2/callback"
+    OAUTH2_PROXY_EMAIL_DOMAIN    = "\"*\"" # oauth proxy will allow all email domains only when the value is "*"
+    OAUTH2_PROXY_OIDC_ISSUER_URL = "https://login.microsoftonline.com/${local.aad_tenant_id}/v2.0"
+    OAUTH2_PROXY_JWKS_ENDPOINT   = "https://login.microsoftonline.com/${local.aad_tenant_id}/discovery/v2.0/keys"
```
Comment (marrobi) [inline templates/workspace_services/guacamole/terraform/web_app.tf:46]: There are instances where these will need to be customized. 

For example if access to the web app is via a custom FQDN, or a different OIDC provider is used. Again might not be one to handle in this PR, but we should create issues to track.

File: docs/tre-admins/setup-instructions/installing-workspace-service-and-user-resource.md [hunk 5]
```diff
@@ -74,7 +74,9 @@ Now that we have published and registered both workspace service and user resour
       "properties": {
         "display_name":"Virtual Desktop",
         "description":"Create virtual desktops for running research workloads",
-        "openid_client_id":"<WORKSPACE_API_CLIENT_ID>",
+        "ws_client_id":"<WORKSPACE_API_CLIENT_ID>",
+        "ws_client_secret":"<WORKSPACE_API_CLIENT_SECRET>",
+        "oauth2_proxy_email_domain": "<email domain>",
```
[the remark this answers is not in the document]
Comment (anatbal) [reply docs/tre-admins/setup-instructions/installing-workspace-service-and-user-resource.md]: This is already marked out, the email domains is now "*"

File: templates/workspace_services/guacamole/guacamole-server/docker/services/oauth/run [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+--upstream http://0.0.0.0:8080  \
+--email-domain "${OAUTH2_PROXY_EMAIL_DOMAIN}"  \
+--redirect-url "${OAUTH2_PROXY_REDIRECT_URI}" --pass-host-header true \
+--show-debug-on-error true --pass-authorization-header true --pass-user-headers true \
+--http-address http://0.0.0.0:8085 \
+--https-address https://0.0.0.0:8086 \
+--cookie-secure false \
+--reverse-proxy true \
+--pass-access-token true \
+--pass-authorization-header true \
```
[the remark this answers is not in the document]
Comment (anatbal) [reply templates/workspace_services/guacamole/guacamole-server/docker/services/oauth/run]: Fixed

File: templates/workspace_services/guacamole/terraform/web_app.tf (hunk 3 above)
Comment (anatbal) [reply templates/workspace_services/guacamole/terraform/web_app.tf:42]: This is the workspace_app_id and the workspace_app_secret that are being transferred from the .env file. So it is not created as part of the workspace deployment...

**Comment by marrobi** [comment]
I'm getting:

```
id_token did not contain an email and profileURL is not defined
```

Any ideas?

**Comment by anatbal** [comment]
> I'm getting:
> 
> ```
> id_token did not contain an email and profileURL is not defined
> ```
> 
> Any ideas?

When are you getting this?

File: templates/workspace_services/guacamole/terraform/web_app.tf (hunk 3 above)
Comment (anatbal) [reply templates/workspace_services/guacamole/terraform/web_app.tf:42]: Do you think it should be part of this PR scope?

Comment (anatbal) [reply templates/workspace_services/guacamole/terraform/web_app.tf:42]: https://github.com/microsoft/AzureTRE/issues/1693
Feel free to change

File: templates/workspace_services/guacamole/guacamole-server/docker/services/oauth/run [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+cookiesecret=$(dd if=/dev/urandom bs=32 count=1 2>/dev/null | base64 | tr -d -- '\n' | tr -- '+/' '-_'; echo)
+
+./oauth2-proxy-v7.2.1.linux-amd64/oauth2-proxy \
+--provider oidc \
+--skip-provider-button \
+--cookie-secret "${cookiesecret}" \
+--oidc-issuer-url "${OAUTH2_PROXY_OIDC_ISSUER_URL}" \
+--upstream http://0.0.0.0:8080  \
+--email-domain "${OAUTH2_PROXY_EMAIL_DOMAIN}"  \
```
Comment (marrobi) [inline templates/workspace_services/guacamole/guacamole-server/docker/services/oauth/run:18]: Is there a reason we are passing values that are already set as environment variables, reading https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/overview#environment-variables there should not be the need?

Comment (eladiw) [reply templates/workspace_services/guacamole/guacamole-server/docker/services/oauth/run:18]: you are right that its redundant. I like to have it explicit, this way you can search it in the code...
and also when you see the command itself you see these params are used, so its more straighforward...

**Review by marrobi — APPROVED** [review]
LGTM

**Pull request merged**